File: examples3d/locked_rotations3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

// This shows a bug when a cylinder is in contact with a very large
// but very thin cuboid. In this case the EPA returns an incorrect
// contact normal, resulting in the cylinder falling through the floor.
pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * The ground
     */
    let ground_size = 5.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height, 0.0));
    let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
    let (_handle, _) = world.insert(rigid_body, collider);

    /*
     * A rectangle that only rotates along the `x` axis.
     */
    let rigid_body = RigidBodyBuilder::dynamic()
        .translation(Vector::new(0.0, 3.0, 0.0))
        .lock_translations()
        .enabled_rotations(true, false, false);
    let collider = ColliderBuilder::cuboid(0.2, 0.6, 2.0);
    let (_handle, _) = world.insert(rigid_body, collider);

    /*
     * A tilted capsule that cannot rotate.
     */
    let rigid_body = RigidBodyBuilder::dynamic()
        .translation(Vector::new(0.0, 5.0, 0.0))
        .rotation(Vector::X * 1.0)
        .lock_rotations();
    let collider = ColliderBuilder::capsule_y(0.6, 0.4);
    let (handle, _) = world.insert(rigid_body, collider);
    let collider = ColliderBuilder::capsule_x(0.6, 0.4);
    world.insert_collider_with_parent(collider, handle);

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec3::new(10.0, 3.0, 0.0), Vec3::new(0.0, 3.0, 0.0));
}

File: python/rapier-testbed/rapier_testbed/examples3/locked_rotations3.py
"""Port of examples3d/locked_rotations3.rs."""
from __future__ import annotations

import rapier3d as rp
from .._registry import register

CATEGORY = "Dynamics"
NAME = "Locked rotations"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 5.0
    ground_height = 0.1
    ground = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_h = bodies.insert(ground)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size), ground_h, bodies
    )

    # Rectangle that only rotates along the x axis.
    body = (
        rp.RigidBody.dynamic()
        .translation((0.0, 3.0, 0.0))
        .locked_axes(rp.LockedAxes.TRANSLATION_LOCKED)
        .enabled_rotations((True, False, False))
    )
    h = bodies.insert(body)
    colliders.insert_with_parent(rp.Collider.cuboid(0.2, 0.6, 2.0), h, bodies)

    # Tilted capsule that cannot rotate.
    body = (
        rp.RigidBody.dynamic()
        .translation((0.0, 5.0, 0.0))
        .rotation((1.0, 0.0, 0.0))
        .locked_axes(rp.LockedAxes.ROTATION_LOCKED)
    )
    h = bodies.insert(body)
    colliders.insert_with_parent(rp.Collider.capsule_y(0.6, 0.4), h, bodies)
    colliders.insert_with_parent(rp.Collider.capsule_x(0.6, 0.4), h, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((10.0, 3.0, 0.0), (0.0, 3.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
